### PR TITLE
feat: multi-currency in Opportunity

### PR DESCRIPTION
**Pull request by anupamvs**
- Made Item table similar to other sales/ purchase Items
- Added conversion rate. The currency field was already there.
- Rate, Amount field in Items table.
- Total fields in the parent form:
   - Total, Base Total: a total of all opportunity items 
   - Grand Total, Base Grand Total: `total` + `opportunity_amount`  or `base_total` + `base_opportunity_amount`

image

`no-docs`

### CODE REVIEW

File: erpnext/crm/doctype/lead/lead.json [hunk 1]
```diff
@@ -454,13 +456,21 @@
...
+   "fieldname": "column_break_22",
+   "fieldtype": "Column Break"
   }
  ],
  "icon": "fa fa-user",
  "idx": 5,
  "image_field": "image",
  "links": [],
- "modified": "2021-01-06 19:39:58.748978",
+ "modified": "2021-05-27 16:42:55.453326",
```
Comment (ruchamahabal) [inline erpnext/crm/doctype/lead/lead.json]: Undo the changes in this file since you have already refactored the lead form in another PR.

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 2]
```diff
@@ -122,10 +122,37 @@ frappe.ui.form.on("Opportunity", {
 		}
 	},
 
+	currency: function(frm) {
```
Comment (ruchamahabal) [inline erpnext/crm/doctype/opportunity/opportunity.js:135]: On changing the currency, labels for all currency fields should show the currency too.

image

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 3]
```diff
@@ -36,6 +36,20 @@ def validate(self):
 		if not self.with_items:
 			self.items = []
 
+		else:
+			self.calculate_totals()
+
+	def calculate_totals(self):
+		total = base_total = 0
+		for item in self.get('items'):
+			item.amount = item.rate * item.qty
```
Comment (ruchamahabal) [inline erpnext/crm/doctype/opportunity/opportunity.py]: image

Keep the qty 1 by default

```python
Traceback (most recent call last):
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/app.py", line 66, in application
    response = frappe.api.handle()
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/handler.py", line 29, in handle
    data = execute_cmd(cmd)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/handler.py", line 65, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/__init__.py", line 1174, in call
    return fn(*args, **newargs)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/desk/form/save.py", line 20, in savedocs
    doc.save()
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 282, in save
    return self._save(*args, **kwargs)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 304, in _save
    self.insert()
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 235, in insert
    self.run_before_save_methods()
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 961, in run_before_save_methods
    self.run_method("validate")
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 855, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 1148, in composer
    return composed(self, method, *args, **kwargs)
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 1131, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File "/Users/ruchamahabal/dev-bench/apps/frappe/frappe/model/document.py", line 849, in <lambda>
    fn = lambda self, *args, **kwargs: getattr(self, method)(*args, **kwargs)
  File "/Users/ruchamahabal/dev-bench/apps/erpnext/erpnext/crm/doctype/opportunity/opportunity.py", line 40, in validate
    self.calculate_totals()
  File "/Users/ruchamahabal/dev-bench/apps/erpnext/erpnext/crm/doctype/opportunity/opportunity.py", line 45, in calculate_totals
    item.amount = item.rate * item.qty
TypeError: unsupported operand type(s) for *: 'int' and 'NoneType'
```

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 4]
```diff
@@ -36,6 +36,20 @@ def validate(self):
...
+			item.basic_rate = self.conversion_rate * item.rate
+			item.base_amount = self.conversion_rate * item.amount
+			total += item.amount
+			base_total += item.base_amount
+		self.total = total
+		self.base_total = base_total
```
Comment (ruchamahabal) [inline erpnext/crm/doctype/opportunity/opportunity.py]: Use `flt`

**Review by ruchamahabal — CHANGES_REQUESTED** [review]

File: erpnext/crm/doctype/lead/lead.json (hunk 1 above)
Comment (anupamvs) [reply erpnext/crm/doctype/lead/lead.json]: Done

File: erpnext/crm/doctype/opportunity/opportunity.json [hunk 5]
```diff
@@ -425,12 +431,50 @@
...
+   "read_only": 1
+  },
+  {
+   "fieldname": "conversion_rate",
+   "fieldtype": "Float",
+   "label": "Exchange Rate"
+  },
+  {
+   "fieldname": "column_break_33",
+   "fieldtype": "Column Break"
```
Comment (ruchamahabal) [inline erpnext/crm/doctype/opportunity/opportunity.json:473]: 1. Add a patch to set the `base_opportunity_amount` from `opportunity_amount`.
2. We should also probably add a Grand Total field which will be Opportunity Amount + Item Total.

**Comment by ruchamahabal** [comment]
Incorrect patch. The currency field was already there in opportunity. 

1. Conversion Rate needs to be set from currency.
2. Set `base_opportunity_amount` from `opportunity_amount` using conversion rate.
3. Set `base_grand_total` equal to `base_opportunity_amount` since item total will be 0 for existing amount as amount field was not there.
4. Set `grand_total` equal to `opportunity_amount` since item total will be 0 for existing amount as amount field was not there.

**Review by ruchamahabal — APPROVED** [review]

**Comment by ruchamahabal** [comment]
No need to backport. It will be a part of v14 only as it's a breaking change to some extent.
